src/lib/storage/segment_iterables/base_segment_iterators.hpp

```diff
@@ -38,7 +38,16 @@ namespace opossum {
  * };
  */
 template <typename Derived, typename Value>
-class BaseSegmentIterator : public boost::iterator_facade<Derived, Value, boost::random_access_traversal_tag, Value> {};
+class BaseSegmentIterator
+    : public boost::iterator_facade<Derived, Value, boost::random_access_traversal_tag, Value, std::ptrdiff_t> {
+ public:
+  // boost will not use the random_access_iterator_tag if reference_type is not a c++ reference (which it isn't here)
+  // we still want to use random access (for binary search, distance, ...)
+  /* typedef std::random_access_iterator_tag iterator_category; */
+  // TODO
+  // this makes some tests segfault. We think it's due to some bad distance_to implementation, but we're not sure.
+  // It doesn't give a noticable performance boost, so leaving it out might also be fine?
```

Comment:
@Bensk1 Boost will not use the random access iterator tag here, even if we pass `random_access_traversal_tag` to the `iterator_facade`, due to the logic documented [here](https://www.boost.org/doc/libs/1_72_0/libs/iterator/doc/iterator_facade.html) (search for "where iterator-category is defined as follows").
However, in our last benchmarks, we couldn't see a difference between adding this here and using the default. (Keep in mind: The AbstractPosList will use `random_access_iterator_tag` no matter what we set here. We just thought it would make sense to also set the random access tag here.)

Thus, we'd suggest that we simply add a comment linking to the boost documentation and explaining that this Iterator will not be treated as a random access iterator by the STL. Does this sound good / should we do something else?

Reply:
It's should be no news to us that boost does not use the random access iterator tag here. During our last meeting we were only discussing the use of `/* typedef std::random_access_iterator_tag iterator_category; */` Which tests are segfaulting? @Bouncner might be interested in that. 

Your proposed solution is alright.

Reply:
Hmm, turns out the current code doesn't segfault anymore when putting the tag in. I guess some changes on master fixed this?

I guess we'll just put in the tag then and remove the TODO?

Reply:
@Bouncner, what was our decision on using `std::random_access_iterator_tag`?

Reply:
Sorry, did not realize that I was tagged here. To be honest, I don't like to have this hack in the master. Is it still true that you cannot get the tests running with the tag?

Could you give me a link to the mentioned `distance` method?

Reply:
And the problem is usually (at least for the iterables currently in the master) not that boost is not using/setting the tag. It actually does. But there is a difference here between boost and the STL tags. The STL talks about random access and has a lot of constraints there, boost talks about random access **traversal**. This traversal works and boost correctly sets the tags for the iterables, but that tag is not something the STL knows or recognizes.

Reply:
> Is it still true that you cannot get the tests running with the tag?

The problem was that we could not get the tests running _with_ the tag.

For the `PosListIterator` class, we want all the STL method that can be implemented more efficiently when random access is available to use that (to be concrete: We needed `std::distance` in O(1)). To get to this, we added the random access tag to the PosListIterator class.

Adding it here is not really in our scope. We just thought that it would make sense to apply this everywhere where we define custom iterators that we will with high probability use `std::distance` on as this could easily cause performance penalties.

We are also fine with undoing all the changes we did here.

Reply:
related distance calls for the PosListIterator are (non-exhaustive):
- src/lib/storage/chunk_encoder.cpp:60
- src/lib/storage/frame_of_reference_segment/frame_of_reference_encoder.hpp:45
- src/lib/storage/segment_iterables/base_segment_iterators.hpp:81
- src/lib/storage/reference_segment/reference_segment_iterable.hpp:151,210
- src/lib/storage/dictionary_segment/dictionary_encoder.hpp:42
- src/lib/storage/lz4_segment/lz4_encoder.hpp:80

Reply:
If it is only about `std::distance()`, we can do it manually `end - it` and add a todo to use `std::distance()` again, once #1531 is resolved.
`end - it` is the same in the end and should directly call the iterator's `difference_to` method.

However, we need to find the issue that the branch fails when the tag is set. Because I hope that we'll fix the iterators one day.

But any decision here is up to @Bensk1 .

Reply:
> However, we need to find the issue that the branch fails when the tag is set. Because I hope that we'll fix the iterators one day.

The segfault with the tag was fixed when we merged master about 2 weeks ago.

> If it is only about `std::distance()`, we can do it manually `end - it` and add a todo to use `std::distance()` again, once #1531 is resolved.
> `end - it` is the same in the end and should directly call the iterator's `difference_to` method.

I'd estimate that to be 12 places in the code then (for my list above, I went through about half the occurences of std::distance in the code). I'm okay with doing that, but need confirmation that we actually want to do it.

Reply:
Do I understand the previous discussion correctly in that we could avoid the `typedef` hack *without losing performance* if we simply replace `std::distance` by `end - it` within the pos list? If yes, by all means, please go for it.

Reply:
From what I understood, the replacements wouldn't be _within_ the pos list but at the locations where iterators on pos lists are used, so its scattered around the code base (see the files I listed above).

Reply:
Bad phrasing, that's what I meant.

Reply:
Alright, will replace by `end - it` and remove the tag.

Reply:
Done. Turns out it's even fewer places than I initially thought.
Should I add a comment saying  that these subtractions should at some point be replaced with `std::distance` again?

Reply:
I don't see any advantage in `std::distance`, so I'd say no